File: Customer/src/main/java/com/postman/alt/service/OrganizationService.java
package com.postman.alt.service;

import com.postman.alt.service.dto.OrganizationInvitationCreateRequest;
import com.postman.alt.service.dto.OrganizationInvitationResponse;
import com.postman.alt.service.dto.OrganizationMemberCreateRequest;
import com.postman.alt.service.dto.OrganizationMemberCreateResponse;
import com.postman.alt.service.dto.OrganizationMemberResponse;

import java.util.List;
import java.util.UUID;

public interface OrganizationService {

    void setProjectCreationAccess(UUID targetUserId, UUID requesterId, boolean canCreateProjects);

    void setMemberManagementAccess(UUID targetUserId, UUID requesterId, boolean canManageMembers);

    List<OrganizationMemberResponse> listMembers(UUID requesterId);

    // creates the AppUser immediately with a system-generated temporary
    // password (mustResetPassword = true), instead of an invitation the
    // person has to redeem themselves - same requireAdmin gate and
    // global-email-uniqueness check as createInvitation.
    OrganizationMemberCreateResponse createMember(UUID requesterId, OrganizationMemberCreateRequest request);

    OrganizationInvitationResponse createInvitation(UUID requesterId, OrganizationInvitationCreateRequest request);

    List<OrganizationInvitationResponse> listInvitations(UUID requesterId);

    void revokeInvitation(UUID invitationId, UUID requesterId);

    // soft-deletes the target user (see AppUser.deletedAt) and removes every
    // one of their project memberships across the whole org - they lose all
    // project/role access immediately and can no longer log in. The owner
    // and the requester themselves can't be deleted this way.
    void deleteMember(UUID targetUserId, UUID requesterId);
}

File: client/src/app/core/services/organization.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { API_BASE_URL } from '../config/api.config';
import { AuthResponse } from '../models/auth.model';
import {
  OrganizationInvitationAcceptRequest,
  OrganizationInvitationCreateRequest,
  OrganizationInvitationResponse,
  OrganizationMemberCreateRequest,
  OrganizationMemberCreateResponse,
  OrganizationMemberResponse,
} from '../models/organization.model';

const BASE = `${API_BASE_URL}/api`;

@Injectable({ providedIn: 'root' })
export class OrganizationService {
  private readonly http = inject(HttpClient);

  listMembers(): Observable<OrganizationMemberResponse[]> {
    return this.http.get<OrganizationMemberResponse[]>(`${BASE}/organizations/members`);
  }

  // creates the account immediately with a system-generated temporary
  // password (returned once in the response) instead of an email invite.
  createMember(request: OrganizationMemberCreateRequest): Observable<OrganizationMemberCreateResponse> {
    return this.http.post<OrganizationMemberCreateResponse>(`${BASE}/organizations/members`, request);
  }

  // removes every one of the user's project memberships and blocks them
  // from logging in again - their existing session is invalidated on the
  // very next request too. The org owner and your own account can't be
  // deleted this way (the backend rejects both).
  deleteMember(userId: string): Observable<void> {
    return this.http.delete<void>(`${BASE}/organizations/members/${userId}`);
  }

  setProjectCreationAccess(userId: string, canCreateProjects: boolean): Observable<void> {
    return this.http.patch<void>(`${BASE}/users/${userId}/project-creation-access`, { canCreateProjects });
  }

  setMemberManagementAccess(userId: string, canManageMembers: boolean): Observable<void> {
    return this.http.patch<void>(`${BASE}/users/${userId}/member-management-access`, { canManageMembers });
  }

  createInvitation(request: OrganizationInvitationCreateRequest): Observable<OrganizationInvitationResponse> {
    return this.http.post<OrganizationInvitationResponse>(`${BASE}/organizations/invitations`, request);
  }

  listInvitations(): Observable<OrganizationInvitationResponse[]> {
    return this.http.get<OrganizationInvitationResponse[]>(`${BASE}/organizations/invitations`);
  }

  revokeInvitation(invitationId: string): Observable<void> {
    return this.http.delete<void>(`${BASE}/organizations/invitations/${invitationId}`);
  }

  acceptInvitation(token: string, request: OrganizationInvitationAcceptRequest): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(`${BASE}/organizations/invitations/${token}/accept`, request);
  }
}
